cmd/kubectl-gadget/snapshot/snapshot.go

```diff
@@ -15,16 +15,118 @@
 package snapshot
 
 import (
-	"github.com/kinvolk/inspektor-gadget/cmd/kubectl-gadget/utils"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"text/tabwriter"
 
 	"github.com/spf13/cobra"
-)
 
-// All the gadgets within this package use this global variable, so let's
-// declare it here.
-var params utils.CommonFlags
+	"github.com/kinvolk/inspektor-gadget/cmd/kubectl-gadget/utils"
+	gadgetv1alpha1 "github.com/kinvolk/inspektor-gadget/pkg/apis/gadget/v1alpha1"
+	processcollectortypes "github.com/kinvolk/inspektor-gadget/pkg/gadgets/process-collector/types"
+	socketcollectortypes "github.com/kinvolk/inspektor-gadget/pkg/gadgets/socket-collector/types"
+	eventtypes "github.com/kinvolk/inspektor-gadget/pkg/types"
+)
 
 var SnapshotCmd = &cobra.Command{
 	Use:   "snapshot",
 	Short: "Take a snapshot of a subsystem and print it",
 }
+
+type GadgetEvent interface {
+	socketcollectortypes.Event | processcollectortypes.Event
+
+	// The Go compiler does not support accessing a struct field x.f where x is
+	// of type parameter type even if all types in the type parameter's type set
+	// have a field f. We may remove this restriction in Go 1.19. See
+	// https://tip.golang.org/doc/go1.18#generics.
+	GetBaseEvent() eventtypes.Event
+}
+
+type GadgetFlags interface {
+	SocketFlags | ProcessFlags
+}
+
+// getSnapshotCallback returns the callback that will be called when a snapshot
+// gadget finishes without errors and generates a list of results per node. This
+// function merges, sorts and print all of them in the requested mode.
+func getSnapshotCallback[Event GadgetEvent, Flags GadgetFlags](
+	outputConf *utils.OutputConfig,
+	gadgetFlags *Flags,
+	sortEvents func([]Event, *Flags),
+	getColsHeader func(*Flags, []string) string,
+	transformEvent func(*Event, *Flags, *utils.OutputConfig) string,
```

Comment:
I am investigating how to create an `interface` with these methods to avoid passing all these arguments but I am not sure it is possible using generics.

Reply:
I found a way to use interfaces, and it improved the code!